### DISCUSSION TITLE
Using examples stored in DocBin in the config to train the NER

### DISCUSSION THREAD

**Discussion by oliviercwa** [explosion/spaCy#9031] [Help: Coding & Implementations] [answered]
In Spacy 3.1.1, I am trying to train a NER with custom objects.

All my examples are stored in a DocBin stored to a file using
```python
    db.to_disk('my_examples')
```

My pipeline is described through a configuration file in which I have:
[initialize.components.ner.labels]
```
[initialize.components.ner.labels]
@readers = "spacy.read_labels.v1"
path = "./my_examples"
require = false
```
After loading the pipeline from the configuration, I call
```python
nlp.initialize()
```

and I am getting an error in read_labels
```
'utf-8' codec can't decode byte 0x9c in position 1: invalid start byte
```

Clearly the format is not accepted. 

How do you express in the configuration that the labels were stored using DocBin ?

Thank you

**Comment by polm** [comment, marked as answer]
Ah, sorry the error there wasn't very helpful. The labels need to be a list of strings saved in a JSON file, you can't provide a DocBin with all the labels. 

We can probably improve that error message.

**Comment by oliviercwa** [comment]
No problems and thanks for the quick answer.

The example for "Preparing Training Data" on this page https://spacy.io/usage/training#training-data shows the use of the DocBin, yet I did not see any way to export from a DocBin to JSON. So is there any advantage to use the DocBin other than storing the examples ?

I just want to make sure I did not miss anything before writing an export function :) especially given the comments on the same page

_"What about the spaCy JSON format...In spaCy v2, the recommended way to store training data was in a particular JSON format, but in v3 this format is deprecated. It’s fine as a readable storage format, but there’s no need to convert your data to JSON before creating a .spacy file."_

and this comment on https://spacy.io/api/data-formats

_"As of v3.0, the JSON input format is deprecated and is replaced by the binary format. Instead of converting Doc objects to JSON, you can now serialize them directly using the DocBin container and then use them as input data."_

**Reply by polm** [reply]
Ah, I see, sorry I didn't spot the issue here. There's a difference between training data and labels. 

The DocBin / `.spacy` files are used for **training data**, which is Docs with annotations. This is sometimes called "labeled data" or even "labels", but it's not the same as the "labels" we're referring to here. You must have training data to train a model.

In the context of `read_labels`, the labels are just the different labels you use in your NER model, like PERSON, ORG, LOC, or whatever. They are not full examples, it's just a list of strings like `["PERSON", "ORG", "LOC", ...]`. 

Normally you don't have to set the labels in advance - the option to do so is only provided for certain specific scenarios, like when you have too much training data to fit in memory. Take a look at the [example projects](https://github.com/explosion/projects/tree/v3/tutorials), where there are several NER ones and none of them use `read_labels`. 

It sounds like maybe you don't need `read_labels` at all, how did you end up adding it to your config?

**Reply by oliviercwa** [reply]
That makes sense. Thank you very much for the explanation.
